bcda/auth/backend.go

```diff
@@ -23,17 +33,40 @@ func init() {
 // Hash is a cryptographically hashed string
 type Hash string
 
-// NewHash creates a hashed string from a source string, return it as a new Hash value.  While this is not
-// intended for production password hashing, as a matter of principle we intend to add a salt and implement
-// bcrypt or another hash appropriate for storing secrets in ticket BCDA-1130.
-func NewHash(source string) Hash {
-	sum := sha256.Sum256([]byte(source))
-	return Hash(fmt.Sprintf("%x", sum))
+// NewHash creates a Hash value from a source string
+// The HashValue consists of the salt and hash separated by a colon ( : )
+// If the source of randomness fails it returns an error.
+func NewHash(source string) (Hash, error) {
+	salt := make([]byte, saltSize)
```

Comment:
If it is invalid to hash "", and it won't match on the hash compare, why hash "" at all?  Can this just short circuit and return "invalid source" error or similar?

Reply:
That's reasonable in most cases.  However, because this is a core function to our authentication, what I'm trying to cover is a future case if misunderstanding, or the case of unexpected failure in our the libraries, or some other failure case that leads to inadvertently passing "" to `pbkdf2.Key()`.  So by definition we can't short-circuit it completely at the input stage.  Where we call the `NewHash()` at the moment, we're providing a random secret anyway.  But I'll add a check here too.  😃